src/Diagnostics.DataProviders/DataProviders/KustoDataProvider.cs

```diff
@@ -4,26 +4,120 @@
 using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Data;
+using System.IO;
+using System.IO.Compression;
 using System.Linq;
+using System.Runtime.CompilerServices;
 using System.Text;
 using System.Threading.Tasks;
+using System.Web;
 
 namespace Diagnostics.DataProviders
 {
+    public class KustoQuery
+    {
+        public string Text;        
+        public string Url;
+    }
     public class KustoDataProvider: DiagnosticDataProvider, IDiagnosticDataProvider
     {
         private KustoDataProviderConfiguration _configuration;
         private IKustoClient _kustoClient;
 
-        public KustoDataProvider(OperationDataCache cache, KustoDataProviderConfiguration configuration): base(cache)
+        public KustoDataProvider(OperationDataCache cache, KustoDataProviderConfiguration configuration) : base(cache)
         {
             _configuration = configuration;
             _kustoClient = KustoClientFactory.GetKustoClient(configuration);
+            Metadata = new DataProviderMetadata
+            {
+                ProviderName = "Kusto"
+            };
         }
 
         public async Task<DataTable> ExecuteQuery(string query, string stampName, string requestId = null, string operationName = null)
         {
+            AddQueryInformationToMetadata(query, stampName);
             return await _kustoClient.ExecuteQueryAsync(query, stampName, requestId, operationName);
         }
+
+        private void AddQueryInformationToMetadata(string query, string stampName)
+        {
+            KustoQuery q = new KustoQuery
+            {
+                Text = query,               
+                Url = CreateKustoQueryUri(stampName, query)
+            };
+
+            Metadata.PropertyBag.Add(new KeyValuePair<string, object>("Query", q));
+        }
+
+        private string ParseRegionFromStamp(string stampName)
+        {
+            if (string.IsNullOrWhiteSpace(stampName))
+            {
+                throw new ArgumentNullException("stampName");
+            }
+
+            var stampParts = stampName.Split(new char[] { '-' });
+            if (stampParts.Any() && stampParts.Length >= 3)
+            {
+                return stampParts[2];
+            }
+
+            //return * for private stamps if no prod stamps are found
+            return "*";
+        }
+
+        private string CreateKustoQueryUri(string stampName, string query)
```

Comment:
Yes, I did mention in my PR, that one method is duplicated and it was this one ParseRegionFromStamp  :-)

Now I tried reusing it but because of the way interface is defined, I wasn't able to call that one and I am open to suggestions on the right way to fix it (I mean, I can just add that to the interface and add a new method to _kustoClient but I dont know if that is the right thing to do) so based on your response, I can make the right change

Reply:
Ya, it doesnt make sense to have to be in the kusto client interface as its not kusto related.. Its more like a helper method which gives you a region based on stamp name. So maybe put this in some reusable helper class that both data provider and client can use.